In [21]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier


In [22]:
df=pd.read_csv("/content/Churn_Modelling.csv")
df.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


In [24]:
df.describe()

RowNumber    CustomerId   CreditScore           Age        Tenure  \
count  10000.00000  1.000000e+04  10000.000000  10000.000000  10000.000000   
mean    5000.50000  1.569094e+07    650.528800     38.921800      5.012800   
std     2886.89568  7.193619e+04     96.653299     10.487806      2.892174   
min        1.00000  1.556570e+07    350.000000     18.000000      0.000000   
25%     2500.75000  1.562853e+07    584.000000     32.000000      3.000000   
50%     5000.50000  1.569074e+07    652.000000     37.000000      5.000000   
75%     7500.25000  1.575323e+07    718.000000     44.000000      7.000000   
max    10000.00000  1.581569e+07    850.000000     92.000000     10.000000   

             Balance  NumOfProducts    HasCrCard  IsActiveMember  \
count   10000.000000   10000.000000  10000.00000    10000.000000   
mean    76485.889288       1.530200      0.70550        0.515100   
std     62397.405202       0.581654      0.45584        0.499797   
min         0.000000       1.000000      0.00000        0.000000   
25%         0.000000       1.000000      0.00000        0.000000   
50%     97198.540000       1.000000      1.00000        1.000000   
75%    127644.240000       2.000000      1.00000        1.000000   
max    250898.090000       4.000000      1.00000        1.000000   

       EstimatedSalary        Exited  
count     10000.000000  10000.000000  
mean     100090.239881      0.203700  
std       57510.492818      0.402769  
min          11.580000      0.000000  
25%       51002.110000      0.000000  
50%      100193.915000      0.000000  
75%      149388.247500      0.000000  
max      199992.480000      1.000000

In [25]:
df.isnull().sum()

RowNumber          0
CustomerId         0
Surname            0
CreditScore        0
Geography          0
Gender             0
Age                0
Tenure             0
Balance            0
NumOfProducts      0
HasCrCard          0
IsActiveMember     0
EstimatedSalary    0
Exited             0
dtype: int64

In [26]:
df.value_counts()

RowNumber  CustomerId  Surname   CreditScore  Geography  Gender  Age  Tenure  Balance    NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  Exited
9984       15656710    Cocci     613          France     Male    40   4       0.00       1              0          0               151325.24        0         1
9983       15768163    Griffin   655          Germany    Female  46   7       137145.12  1              1          0               115146.40        1         1
9982       15672754    Burbidge  498          Germany    Male    42   3       152039.70  1              1          1               53445.17         1         1
9981       15719276    T'ao      741          Spain      Male    35   6       74371.49   1              0          0               99595.67         0         1
9980       15692664    Diribe    677          France     Female  58   1       90022.85   1              0          1               2988.28          0         1
                                                                                                                                                             ..
5          15737888    Mitchell  850          Spain      Female  43   2       125510.82  1              1          1               79084.10         0         1
4          15701354    Boni      699          France     Female  39   1       0.00       2              0          0               93826.63         0         1
3          15619304    Onio      502          France     Female  42   8       159660.80  3              1          0               113931.57        1         1
2          15647311    Hill      608          Spain      Female  41   1       83807.86   1              0          1               112542.58        0         1
1          15634602    Hargrave  619          France     Female  42   2       0.00       1              1          1               101348.88        1         1
Name: count, Length: 10000, dtype: int64

In [27]:
df.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [28]:
#handling missing values
df.drop(["RowNumber", "CustomerId"], axis=1, inplace=True)

**Encoding Categorical Columns**

In [29]:
#Convert categorical variables → numeric (one-hot encoding)
df = pd.get_dummies(df, columns=["Geography", "Gender"], drop_first=True)

**Splitting Feature and Target**

In [31]:
X = df.drop("Exited", axis=1)
y = df["Exited"]

In [34]:
df.drop(["Surname"], axis=1, inplace=True)


In [36]:
X = df.drop("Exited", axis=1)
y = df["Exited"]


**Scaling Features**

In [37]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)


**Train-Test Split**

In [38]:
X_train, X_test, y_train, y_test = train_test_split(
    X_scaled, y, test_size=0.3, random_state=42
)


**Training Logistic Regression**

In [39]:
lr = LogisticRegression()
lr.fit(X_train, y_train)
y_pred_lr = lr.predict(X_test)

print("Logistic Regression Accuracy:", accuracy_score(y_test, y_pred_lr))
print(classification_report(y_test, y_pred_lr))


Logistic Regression Accuracy: 0.8113333333333334
              precision    recall  f1-score   support

           0       0.83      0.96      0.89      2416
           1       0.54      0.20      0.29       584

    accuracy                           0.81      3000
   macro avg       0.69      0.58      0.59      3000
weighted avg       0.78      0.81      0.77      3000



**Accuracy: 81.13%**

But accuracy is misleading because churn data is imbalanced (few churners).

**Class 0 (Not churn)**

**Precision: 0.83**

**Recall: 0.96**

**F1-score: 0.89**

The model is very good at predicting people who stay.

**Class 1 (Churn)**

**Precision: 0.54**

**Recall: 0.20**

**F1-score: 0.29**

**Meaning:**
-The model fails to detect customers who actually churn.
-It misses 80% of churners.

**Evaluation (Accuracy, Precision, Recall, F1)**

In [43]:
# MAE & MSE
from sklearn.metrics import mean_absolute_error, mean_squared_error
print("MAE:", mean_absolute_error(y_test, y_pred_lr))
print("MSE:", mean_squared_error(y_test, y_pred_lr))
print("RMSE:",np.sqrt(mean_squared_error(y_test, y_pred_lr)))

MAE: 0.18866666666666668
MSE: 0.18866666666666668
RMSE: 0.4343577634469847


**Training Random Forest**

In [40]:
rf = RandomForestClassifier(n_estimators=200, random_state=42)
rf.fit(X_train, y_train)
y_pred_rf = rf.predict(X_test)

print("Random Forest Accuracy:", accuracy_score(y_test, y_pred_rf))
print(classification_report(y_test, y_pred_rf))


Random Forest Accuracy: 0.8663333333333333
              precision    recall  f1-score   support

           0       0.88      0.96      0.92      2416
           1       0.75      0.47      0.58       584

    accuracy                           0.87      3000
   macro avg       0.82      0.72      0.75      3000
weighted avg       0.86      0.87      0.85      3000



**Random Forest Accuracy: 0.8663 (≈ 86.6%)**

**Class 0 (Not Churn):**

**Precision 0.88**
When the model predicts Not Churn, it's correct 88% of the time.

**Recall 0.96**
It successfully identifies 96% of all Not Churn transactions.

**F1 0.92**
Strong overall performance on the majority class.

**Class 1 (Churn):**

**Precision 0.75**
When the model predicts churn, it's correct 75% of the time.

**Recall 0.47**
The model detects only 47% of actual churn → this is too low.

**F1 0.58**
Weak performance on the minority (Churn) class

In [41]:
gb = GradientBoostingClassifier()
gb.fit(X_train, y_train)
y_pred_gb = gb.predict(X_test)

print("Gradient Boosting Accuracy:", accuracy_score(y_test, y_pred_gb))
print(classification_report(y_test, y_pred_gb))


Gradient Boosting Accuracy: 0.871
              precision    recall  f1-score   support

           0       0.89      0.96      0.92      2416
           1       0.77      0.48      0.59       584

    accuracy                           0.87      3000
   macro avg       0.83      0.72      0.76      3000
weighted avg       0.86      0.87      0.86      3000



**The Gradient Boosting model gives 87% accuracy**

**(Class 0)**

**Recall = 96%**
The model correctly identifies most customers who continue their subscription.

**(Class 1)**

**Recall = 48%**
The model finds almost half of the customers who are likely to leave.

**Precision = 77%**
When it predicts churn, it is usually correct.